features/netsocket/NetworkStack.h

```diff
@@ -63,13 +62,56 @@ class NetworkStack
     virtual nsapi_error_t gethostbyname(const char *host,
             SocketAddress *address, nsapi_version_t version = NSAPI_UNSPEC);
 
+    /** Hostname translation callback (asynchronous)
+     *
+     *  Callback will be called after DNS resolution completes or a failure
+     *  occurs.
+     *
+     *  @param status  0 on success, negative error code on failure
+     *  @param address On success, destination for the host SocketAddress
+     *  @param data    Caller defined data
+     */
+    typedef mbed::Callback<void (nsapi_error_t result, SocketAddress *address, void *data)> hostbyname_cb_t;
```

Comment:
As @kjbracey-arm mentioned elsewhere, the documentation of callback should state what can be done from it. I guess typical application would want to connect a socket on success case, or issue another DNS query on failure.

Reply:
I'll add a note about limitations. Since this is always running from thread it is possible to do more than e.g. for socket sigio(). But thread stack size limits what can be initiated. 

If stack sizes are configured larger (lwip, nanostack or global event queue thread) or if application provides callback using nsapi_dns_call_in_set(), at least non-blocking UDP operations and asynchronous DNS query are possible from callback.